File: Python/Medium/2075 Decode the Slanted Ciphertext.py
class Solution:
    def decodeCiphertext(self, encodedText: str, rows: int) -> str:
        if not encodedText:
            return ""
        
        n = len(encodedText)
        cols = n // rows
        res = []
        
        # We start each diagonal from the first row (row 0)
        # and move shifted to the right by 'i'
        for i in range(cols):
            r, c = 0, i
            while r < rows and c < cols:
                res.append(encodedText[r * cols + c])
                r += 1
                c += 1
        
        # Join and trim trailing spaces only
        return "".join(res).rstrip()

File: Java/Medium/2075 Decode the Slanted Ciphertext.java
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (encodedText.length() == 0) return "";
        
        int n = encodedText.length();
        int cols = n / rows;
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < cols; i++) {
            for (int r = 0, c = i; r < rows && c < cols; r++, c++) {
                sb.append(encodedText.charAt(r * cols + c));
            }
        }
        
        // Trim only trailing spaces
        int last = sb.length() - 1;
        while (last >= 0 && sb.charAt(last) == ' ') {
            last--;
        }
        return sb.substring(0, last + 1);
    }
}
